README.md
# Logger

Simple lightweight API that accepts a POST request with a JSON payload that is
written to a buffer and flushed at configurable intervals into Elasticsearch.
Whichever setting (time, buffer size) that comes first for the processor
configuration will execute the flush. You can start the entire service with
`make up` to run Elasticsearch and Kibana and then `make run` to run the API.

[![CircleCI](https://circleci.com/gh/cachelab/logger.svg?style=svg)](https://circleci.com/gh/cachelab/logger)

## Usage

This task is configured by the following environment variables:

```bash
FLUSH_INTERVAL    # the number of seconds you want the buffer to be flushed
WORKERS           # the number of workers that will be available to buffer requests
BULK_ACTIONS      # the maximum number of records that will be flushed
MAX_RETRIES       # how many times the client will try to connect to Elasticsearch
ELASTICSEARCH_URL # the url to the Elasticsearch cluster
RUN_ONCE          # used for unit testing to not start the http server
```

## Example

```
curl -XPOST http://127.0.0.1:3000/your-api/log -d '{"message": "Error in line 128. Bad object.", "level": "error", "data": {"http_status": 500}}}'
```

![alt text](/images/screenshot.png)

## Contributing

* `make` - defaults to make build
* `make up` - runs elasticsearch and kibana in the foreground with docker compose
* `make run` - runs the api in a docker container
* `make build` - builds your logger docker container
* `make vet` - go fmt and vet code
* `make test` - run unit tests

Before you submit a pull request please update the semantic version inside of
`main.go` with what you feel is appropriate and then edit the `CHANGELOG.md` with
your changes and follow a similar structure to what is there.


CHANGELOG.md
## 1.2.0 (December 29 2021)

* Updated Go and packages.
* Converted to multi stage build for smaller container size.

## 1.1.1 (November 19 2019)

* Updated Go version for CircleCi to match container.
* Various other cleanup items.

## 1.1.0 (November 11 2019)

* Updated Go docker version.

## 1.0.0 (August 21 2019)

* Initial release.


service_test.go
package main

import (
	"context"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"

	"github.com/olivere/elastic/v7"
	"github.com/stretchr/testify/assert"
)

func TestHttpServer(t *testing.T) {
	client, _ := elastic.NewClient()

	svc := &Service{}
	processor, _ := client.BulkProcessor().Name("processor").Do(context.Background())
	svc.processor = processor

	payload := `{ "message" : "test message", "level" : "error" }`
	body := strings.NewReader(payload)

	req := httptest.NewRequest("POST", "/logs/misc", body)
	w := httptest.NewRecorder()
	svc.handler(w, req)

	resp := w.Result()
	assert.Equal(t, resp.StatusCode, http.StatusNoContent)

	req = httptest.NewRequest("POST", "/", body)
	w = httptest.NewRecorder()
	svc.handler(w, req)

	resp = w.Result()
	assert.Equal(t, resp.StatusCode, http.StatusBadRequest)

	req = httptest.NewRequest("GET", "/logs/misc", nil)
	w = httptest.NewRecorder()
	svc.handler(w, req)

	resp = w.Result()
	assert.Equal(t, resp.StatusCode, http.StatusMethodNotAllowed)

	payload = `fail`
	body = strings.NewReader(payload)

	req = httptest.NewRequest("POST", "/logs/misc", nil)
	w = httptest.NewRecorder()
	svc.handler(w, req)

	resp = w.Result()
	assert.Equal(t, resp.StatusCode, http.StatusBadRequest)

	payload = `{ "message" : "test message", "level" : "error" }`
	body = strings.NewReader(payload)

	req = httptest.NewRequest("POST", "/logs/misc/bad/url", body)
	w = httptest.NewRecorder()
	svc.handler(w, req)

	resp = w.Result()
	assert.Equal(t, resp.StatusCode, http.StatusBadRequest)
}

func TestInit(t *testing.T) {
	svc := Service{}

	err := svc.Init()
	assert.Equal(t, err, nil)

	os.Setenv("FLUSH_INTERVAL", "1")
	os.Setenv("WORKERS", "1")
	os.Setenv("BULK_ACTIONS", "1")
	os.Setenv("MAX_RETRIES", "5")
	os.Setenv("ELASTICSEARCH_URL", "http://127.0.0.1:9200")
	os.Setenv("RUN_ONCE", "true")

	err = svc.Init()
	assert.Equal(t, err, nil)

	os.Setenv("FLUSH_INTERVAL", "fail")

	err = svc.Init()
	assert.NotEqual(t, err, nil)

	os.Setenv("FLUSH_INTERVAL", "1")
	os.Setenv("WORKERS", "fail")

	err = svc.Init()
	assert.NotEqual(t, err, nil)

	os.Setenv("FLUSH_INTERVAL", "1")
	os.Setenv("WORKERS", "1")
	os.Setenv("BULK_ACTIONS", "fail")

	err = svc.Init()
	assert.NotEqual(t, err, nil)

	os.Setenv("FLUSH_INTERVAL", "1")
	os.Setenv("WORKERS", "1")
	os.Setenv("BULK_ACTIONS", "1")
	os.Setenv("MAX_RETRIES", "fail")

	err = svc.Init()
	assert.NotEqual(t, err, nil)

	os.Setenv("FLUSH_INTERVAL", "1")
	os.Setenv("WORKERS", "1")
	os.Setenv("BULK_ACTIONS", "1")
	os.Setenv("MAX_RETRIES", "1")
	os.Setenv("RUN_ONCE", "fail")

	err = svc.Init()
	assert.NotEqual(t, err, nil)
}


service.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/olivere/elastic/v7"
)

// Defaults that can be overridden.
const DefaultFlushInterval = 15 * time.Second
const DefaultWorkers = 1
const DefaultBulkActions = 1000
const DefaultMaxRetries = 5
const DefaultElasticsearchURL = "http://127.0.0.1:9200"
const DefaultRunOnce = false

type Service struct {
	processor *elastic.BulkProcessor
}

type Doc struct {
	Message string                  `json:"message"`
	Level   string                  `json:"level"`
	Date    string                  `json:"date"`
	Data    *map[string]interface{} `json:"data"`
}

func (svc *Service) Init() error {
	var err error
	var flushInterval time.Duration
	var workers int
	var bulkActions int
	var maxRetries int
	var elasticsearchUrl string
	var runOnce bool

	// Check if there is a FLUSH_INTERVAL passed in.
	if os.Getenv("FLUSH_INTERVAL") == "" {
		flushInterval = DefaultFlushInterval
	} else {
		i, err := strconv.Atoi(os.Getenv("FLUSH_INTERVAL"))
		if err != nil {
			return err
		}

		flushInterval = time.Duration(i) * time.Second
	}

	// Check if there is a WORKERS passed in.
	if os.Getenv("WORKERS") == "" {
		workers = DefaultWorkers
	} else {
		workers, err = strconv.Atoi(os.Getenv("WORKERS"))
		if err != nil {
			return err
		}
	}

	// Check if there is a BULK_ACTIONS passed in.
	if os.Getenv("BULK_ACTIONS") == "" {
		bulkActions = DefaultBulkActions
	} else {
		bulkActions, err = strconv.Atoi(os.Getenv("BULK_ACTIONS"))
		if err != nil {
			return err
		}
	}

	// Check for the MAX_RETRIES
	if os.Getenv("MAX_RETRIES") == "" {
		maxRetries = DefaultMaxRetries
	} else {
		maxRetries, err = strconv.Atoi(os.Getenv("MAX_RETRIES"))
		if err != nil {
			return err
		}
	}

	// Check for the ELASTICSEARCH_URL
	if os.Getenv("ELASTICSEARCH_URL") == "" {
		elasticsearchUrl = DefaultElasticsearchURL
	} else {
		elasticsearchUrl = os.Getenv("ELASTICSEARCH_URL")
	}

	// Check for the RUN_ONCE
	if os.Getenv("RUN_ONCE") == "" {
		runOnce = DefaultRunOnce
	} else {
		runOnce, err = strconv.ParseBool(os.Getenv("RUN_ONCE"))
		if err != nil {
			return err
		}
	}

	// Setup the elastic client.
	client, err := elastic.NewClient(
		elastic.SetURL(elasticsearchUrl),
		elastic.SetSniff(false),
		elastic.SetMaxRetries(maxRetries),
	)
	if err != nil {
		return err
	}

	// Setup the bulk processor to flush the logs to elastic.
	processor, err := client.BulkProcessor().
		Name("processor").
		Workers(workers).
		BulkActions(bulkActions).
		BulkSize(2 << 20).
		FlushInterval(flushInterval).
		Do(context.Background())
	if err != nil {
		return err
	}
	defer processor.Close()

	// Setup the service processor.
	svc.processor = processor

	// Setup http server.
	mux := http.NewServeMux()

	// Setup the http handle for incoming logs.
	mux.HandleFunc("/", svc.handler)
	mux.HandleFunc("/ping", svc.ping)

	// Run once mode for unit tests.
	if runOnce {
		return nil
	}

	return http.ListenAndServe(":3000", mux)
}

// Private

func (svc *Service) ping(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	w.WriteHeader(http.StatusOK)
}

func (svc *Service) handler(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	if r.Method == "POST" {
		// Decode the incoming payload that will be placed into the bulker.
		decoder := json.NewDecoder(r.Body)

		var doc Doc
		err := decoder.Decode(&doc)
		if err != nil {
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		// Break up the path to get the index.
		path := strings.Split(r.URL.Path, "/")

		// Must include a path which will be the index and type you are targeting.
		if len(path) != 3 {
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		t := time.Now()

		doc.Date = t.Format(time.RFC3339)

		r := elastic.NewBulkIndexRequest().Index(fmt.Sprintf(`%s-%s`, path[1], t.Format("2006-01-02"))).Type(path[2]).Doc(doc)
		svc.processor.Add(r)

		w.WriteHeader(http.StatusNoContent)
		return
	}

	w.WriteHeader(http.StatusMethodNotAllowed)
}
